public/app/plugins/datasource/grafana-azure-monitor-datasource/resourcePicker/resourcePickerData.ts

```diff
@@ -25,12 +27,86 @@ const RESOURCE_GRAPH_URL = '/providers/Microsoft.ResourceGraph/resources?api-ver
 
 export default class ResourcePickerData extends DataSourceWithBackend<AzureMonitorQuery, AzureDataSourceJsonData> {
   private resourcePath: string;
+  private supportedMetricNamespaces: string[];
 
   constructor(instanceSettings: DataSourceInstanceSettings<AzureDataSourceJsonData>) {
     super(instanceSettings);
     this.resourcePath = `${routeNames.resourceGraph}`;
+    const cloud = getAzureCloud(instanceSettings);
+    this.supportedMetricNamespaces = new SupportedNamespaces(cloud).get();
   }
 
+  async fetchInitialRows(currentSelection: string): Promise<ResourceRowGroup> {
+    const subscriptions = await this.getSubscriptions();
+    if (!currentSelection) {
+      return subscriptions;
+    }
+
+    let resources = subscriptions;
+    const parsedURI = parseResourceURI(currentSelection);
+    if (parsedURI) {
+      const resourceGroupURI = `/subscriptions/${parsedURI.subscriptionID}/resourceGroups/${parsedURI.resourceGroup}`;
+
+      if (parsedURI.resourceGroup) {
+        const resourceGroups = await this.getResourceGroupsBySubscriptionId(parsedURI.subscriptionID);
+        resources = addResources(resources, `/subscriptions/${parsedURI.subscriptionID}`, resourceGroups);
+      }
+
+      if (parsedURI.resource) {
+        const resourcesForResourceGroup = await this.getResourcesForResourceGroup(resourceGroupURI);
+        resources = addResources(resources, resourceGroupURI, resourcesForResourceGroup);
+      }
+    }
+    return resources;
+  }
+
+  async fetchNestedRowData(resourceGroupOrSubscription: ResourceRow) {
+    return resourceGroupOrSubscription.type === ResourceRowType.Subscription
+      ? await this.getResourceGroupsBySubscriptionId(resourceGroupOrSubscription.id)
+      : await this.getResourcesForResourceGroup(resourceGroupOrSubscription.id);
+  }
```

Comment:
nit: this method is making it more difficult to know what's happening. Since it's only used once I would avoid it. Also, let's avoid the name ` resourceGroupOrSubscription` so we don't need to add more `Or...` in the future :sweat_smile:

Reply:
my reason for moving it in the resourcePickerData was not because I foresee us using it again in some other way (none of the methods in here are used all that frequently), but because I think its helpful to separate out our code by functionality, we have a resourcePickerData file, that I think is responsible for fetching and parsing data, and a resourcePicker react component that is responsible for rendering that parsed data and handling user interactions.

I think having everything in one pr makes it difficult to see :D so I took your advice and tried to move this sort of thing in a separate pr: https://github.com/grafana/grafana/pull/48312 I also changed the argument names rather than just copy paste what we had before, let me know what you think if you're interested!